lib/http_server.ex

```diff
@@ -15,19 +18,90 @@ defmodule HTTPServer do
   end
 
   defp serve(socket) do
+    request =
+      socket
+      |> read_line()
+      |> parse_request()
+
+    response =
+      request
+      |> build_response()
+      |> format_response()
+
     socket
-    |> read_line()
-    |> write_line(socket)
+    |> write_line(response)
     |> serve()
   end
 
   defp read_line(socket) do
-    {:ok, data} = :gen_tcp.recv(socket, 0)
-    data
+    case :gen_tcp.recv(socket, 0) do
+      {:ok, data} ->
+        data
+
+      {:error, :econnrefused} ->
+        exit(:normal)
+
+      {:error, :closed} ->
+        exit(:normal)
+    end
   end
 
-  defp write_line(data, socket) do
+  defp write_line(socket, data) do
     :gen_tcp.send(socket, data)
     socket
   end
+
+  def parse_request(message) do
+    [request_data | body] = message |> String.split("\r\n\r\n")
```

Comment:
👀 you might find you'll want a `helpers` or `constants` module for `carriage_return` since you're using it in multiple places/functions (I see you already extracted it below - nice!)